### PR TITLE
fix: SWC plugin

### PR DESCRIPTION
**Pull request by YassinEldeeb**
## Fix reproductions and close issues

- [ ] https://github.com/awinograd/graphql-codegen-client-preset-swc-oom
- [x] https://github.com/dotansimha/graphql-code-generator/issues/9145
- [x] https://github.com/dotansimha/graphql-code-generator/issues/9195
- [ ] https://github.com/dotansimha/graphql-code-generator/issues/9337
- [x] https://github.com/dotansimha/graphql-code-generator/issues/9445
- [ ] https://github.com/huv1k/nextjs-auth-prisma
- [x] https://github.com/dotansimha/graphql-code-generator/issues/9450
- [ ] https://github.com/dotansimha/graphql-code-generator/issues/9457

### CODE REVIEW

**Comment by YassinEldeeb** [comment]
oops, based off of the wrong branch

**Comment by focux** [comment]
Is there any progress on this?

**Comment by saihaj** [comment]
@YassinEldeeb let's rebase this one now?

**Comment by CaptainN** [comment]
I recreated this on top of master, and updated to swc_core 0.79.x but was unable to produce a binary that didn't panic.

```
info Creating an optimized production build .thread 'thread '<unnamed><unnamed>' panicked at '' panicked at 'failed to invoke plugin: failed to invoke plugin on Some(... every file)
```

I don't know enough about swc or rust to go much further.

**Comment by MatteoGiacomo** [comment]
Ehy guys, do you have any updates on this? @YassinEldeeb  @saihaj, thanks

**Comment by YassinEldeeb** [comment]
## SWC Plugin Concerns & Clarifications:

A while ago, I discussed with the team the SWC plugin and the challenges with its integration. Here's an overview of our findings:

#### 1. SWC Core Versioning and Compatibility:

- On mid-may, I discovered that SWC provided documentation on selecting the correct SWC Core version. Essentially, each minor version supports a range of nextjs versions, specifically `@swc/core` - the Next.js compiler.
    
- An important note: Wasm plugins aren't backward compatible. Hence, selecting the correct `swc_core` version for your plugin is crucial.
> From SWC Documentation: "Currently, the Wasm plugins are not backwards compatible. So you need to select an appropriate version of swc_core for your plugin."
    
[LINK to the documentation](https://swc.rs/docs/plugin/selecting-swc-core)


#### 2. My Initial Approach:

- I explored creating an installation CLI. It would read the nextjs version in a project’s package.json and determine the right SWC plugin version to use.
- When publishing from our end, the SWC plugin would provide versions like:
    - `@graphql-codegen/swc-plugin@v0.78.x`
    - `@graphql-codegen/swc-plugin@v0.76.x`
    - `@graphql-codegen/swc-plugin@v0.75.x`
 
And the CLI would decide on the installation version based on the read nextjs version from the project. And, of course, upgrading your nextjs version would likely break the SWC plugin silently.

#### 3. Renovate Configuration Issues:
- Our existing renovate config, which issues PRs to regularly update the `swc_core`, seems obsolete. It tends to favor the latest or even canary versions of Next.js, and that's what introduced the breaking changes from the beginning until now with the SWC plugin not working unless with a specific range of Next.js versions which is specified in their documentation. 
- A potential solution could be to have it auto-publish the plugin with the upcoming SWC core version. However, users must be aware of the need to pin the `swc_core` version. If they update their Next.js version, the SWC plugin also needs an update. The error messages, as you've all seen, might not be straightforward.

#### 4. Future Challenges:
- Keeping up with the frequent releases of SWC Core and Next.js versions is challenging. One idea might be to automate some build and publishing steps based on scraping their documentation, but there's no guarantee it'll always be updated promptly.

#### 5. Development Concerns for SWC Plugin:
- Currently, `swc_core` is developed on the Rust nightly version, and not the `stable` toolchain. We've had to adapt to that environment.

Hoping this sheds some light on the SWC plugin challenges. Feedback and suggestions are welcome!
And I apologize for the long wait on this.

We've been busy preparing for the first [GraphQL Conf 2023 held in San Francisco](https://graphql.org/conf/), and we plan to connect with SWC maintainers and figure out a solution together once we get back.

**Comment by victorandree** [comment]
@YassinEldeeb I appreciate the update! I tried to follow your steps on how to update this on my own – but it fails in different ways (basically confirming what @CaptainN commented above).

Leaving this comment so maybe the community can figure this out.

- If I specify `swc_core@0.79.*`, which should work with `next@v13.4.10-canary.1 ~` (according to https://swc.rs/docs/plugin/selecting-swc-core#v079x), I can build the plugin. However, if I try to then _use_ this plugin with `next@13.4.13` (latest, as of today) or `next@13.4.10-canary.1`, I get errors about "out of bounds memory access" (full stack trace below).
- Note: If I just use the current plugin with `next@13.4.13`, the error is instead `RuntimeError: unreachable`.
- If I simply try to build the current version of the `swc-plugin` (which specifies `swc_core@0.75.*`), I get a build error from `proc-macro2` about "unknown feature `proc_macro_span_shrink`". I'm guessing this is because we have to build this with the Rust nightly toolchain, as mentioned in the comment by @YassinEldeeb, and something has changed here.

Stack trace from `swc_core@0.79.*` with `next@13.4.13`. Node version is 18.17.1 (latest LTS).

```plain
Error: failed to process failed to invoke plugin: failed to invoke plugin on 'Some("/c/next-app/pages/index.tsx")'

Caused by:
    0: failed to invoke `/c/next-app/node_modules/@graphql-codegen/client-preset-swc-plugin/swc_plugin.wasm` as js transform plugin at /c/next-app/node_modules/@graphql-codegen/client-preset-swc-plugin/swc_plugin.wasm
    1: RuntimeError: out of bounds memory access

Stack backtrace:
   0: _napi_register_module_v1
   1: _napi_register_module_v1
   2: _BrotliDecoderVersion
   3: _BrotliDecoderVersion
   4: _BrotliDecoderVersion
   5: _BrotliDecoderVersion
   6: _BrotliDecoderVersion
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: <unknown>
  15: <unknown>
  16: <unknown>
  17: <unknown>

Stack backtrace:
   0: _napi_register_module_v1
   1: _napi_register_module_v1
   2: <unknown>
   3: <unknown>

Import trace for requested module:
./pages/index.tsx
```

**Comment by victorandree** [comment]
Hopefully relevant: The build failure is because of https://github.com/swc-project/swc/issues/7479. Setting `lto = false` for the release profile allows me to build and use a new version of the plugin 🎉.

**Comment by JamesCodes** [comment]
> Hopefully relevant: The build failure is because of [swc-project/swc#7479](https://github.com/swc-project/swc/issues/7479). Setting `lto = false` for the release profile allows me to build and use a new version of the plugin 🎉.

Thanks you! This saved me.

**Comment by YassinEldeeb** [comment]
That's absolutely amazing, @victorandree!! Sorry, I was late getting back to you,
I'm super glad you were able to figure out the issue! ❤️

**Comment by bvobart** [comment]
Hi @YassinEldeeb and @victorandree, thank you both for this MR and your patched NPM package!

This has fixed the original syntax error that I was getting in #9337. Without any further updates to the repo, I was getting an out of bounds memory access error like in https://github.com/dotansimha/graphql-code-generator/pull/9489#issuecomment-1672764552, but that was fixed by upgrading `@swc/core` to `1.3.80` as prescribed by the [SWC docs](https://swc.rs/docs/plugin/selecting-swc-core).

However, the issue isn't entirely fixed yet. As soon as I import code from another TypeScript file, I get a different syntax error, even though I've updated my [example repo](https://github.com/bvobart/gql-client-preset-swc-jest-issue) for #9337 to use the patched SWC plugin and I've updated all dependencies to versions that should be compatible (in particular `@swc/core`).

Could you look into this again please? :blush:

**Comment by victorandree** [comment]
FWIW, I recently noticed that my Next.js builds started ballooning in size, so I wanted to have another look at this issue (suspecting that maybe the plugin broke again). While I haven't had any issues using the plugin with Next.js 14.0.2, I figured it's good to be up-to-date.

TL;DR: New version of my fork released as `@victorandree/graphql-codegen-client-preset-swc-plugin@0.3.0`, using `swc_core@0.86`. Updating `swc_core` required me to change test code; see my fork at https://github.com/victorandree/graphql-code-generator. Unfortunately, did not help with bundle size.

- @YassinEldeeb Now that GraphQL Conf 2023 is over, have you had the chance to talk to the SWC maintainers about a way forward?
- @bvobart Unfortunately, I'm not familiar with using `@swc/jest` to transform files for jest. I tried running your reproduction repository, and for me it fails even if I remove the plugin. At a glance, it seems related to your TypeScript code not being compiled - because the syntax error is related to `a: number` in `Foo.ts`.

IMHO, the concern about matching `swc_core` and Next.js versions is valid, but we're also all consenting adults here, who are using the known unstable `swcPlugins` feature and are seeing "Experiments (use at your own risk): swcPlugins" on every build with [a big warning about it in the docs](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#swc-plugin). 

For now, I think it'd be worthwhile to release a new version of the plugin with an updated `swc_core`, considering it's been broken at a patch version of Next.js for some time now.

**Comment by YassinEldeeb** [comment]
Hey, really sorry for the very late reply! We are currently tracking the issue with SWC maintainers to figure out a solution.

Please add a 👍 to the issue and feel free to join the conversation.
https://github.com/swc-project/swc/issues/8315

I'll keep you updated!

**Comment by YassinEldeeb** [comment]
@victorandree @bvobart  @MatteoGiacomo  @CaptainN @focux  as you can see, unfortunately, they replied simply with `In short, this is impossible`, please refer to the above issue's comment from the maintainers.

**Comment by victorandree** [comment]
@YassinEldeeb I appreciate trying to sync with the SWC maintainers on this, and I also appreciate the challenges of trying to keep things in sync between SWC/Next.js versions (and that Next.js isn't the only game in town).

That said, would you consider releasing new versions of the plugin with documentation to clarify compatibility, or is this effectively unmaintained until SWC stabilizes?

**Comment by YassinEldeeb** [comment]
@victorandree I really appreciate your enthusiasm and engagement in this issue!
> would you consider releasing new versions of the plugin with documentation to clarify compatibility

Unfortunately, SWC maintainers weren't able to promise us with up to date documentation on the compatibility between next.js versions and rust's `swc_core` crate (package), which is the main concern causing all of the issues we're all currently facing.

> is this effectively unmaintained until SWC stabilizes?

It's difficult, but I wouldn't say it's completely unmaintained, it's just about chasing up to date comparability information from SWC, they have effectively claimed plugins to be a "Beta Feature", to avoid people complaining about them releasing breaking changes without notices. So it's tricky.

We are thinking of moving the SWC plugin to the [GraphQL Codegen Community](https://github.com/dotansimha/graphql-code-generator-community) repo, where the community can file issues of breaking changes as they detect it, lookup https://swc.rs/docs/plugin/selecting-swc-core for updated documentation, if not, ask SWC maintainer to update it by filing an issue, then finally releasing a new version of the SWC plugin, so there would be:
- @graphql-codegen/swc-plugin@v0.78.x
- @graphql-codegen/swc-plugin@v0.76.x
- @graphql-codegen/swc-plugin@v0.75.x

Where each new version doesn't override the previous one, but adds support for newer next.js versions.

So, what do you think @victorandree? Would it be feasible to let the community take care of maintaining it given the challenges mentioned?

Curious to hear your thoughts! 👀

**Comment by victorandree** [comment]
@YassinEldeeb

> We are thinking of moving the SWC plugin to the [GraphQL Codegen Community](https://github.com/dotansimha/graphql-code-generator-community) repo, where the community can file issues of breaking changes as they detect it, lookup https://swc.rs/docs/plugin/selecting-swc-core for updated documentation, if not, ask SWC maintainer to update it by filing an issue, then finally releasing a new version of the SWC plugin, so there would be:
> 
> * `@graphql-codegen/swc-plugin@v0.78.x`
> * `@graphql-codegen/swc-plugin@v0.76.x`
> * `@graphql-codegen/swc-plugin@v0.75.x`
> 
> Where each new version doesn't override the previous one, but adds support for newer next.js versions.
> 
> So, what do you think @victorandree? Would it be feasible to let the community (including most people in the current thread) take care of maintaining it given the challenges mentioned?

- I definitely hope that the community can help with maintaining this, although I'm not sure what difference it makes if this is part of [dotansimha/graphql-code-generator-community](https://github.com/dotansimha/graphql-code-generator-community) or [dotansimha/graphql-code-generator](https://github.com/dotansimha/graphql-code-generator). The contribution guidelines seem identical for both. Maybe I'm just not up-to-date with your governance structure :)
- I'm not super convinced on tracking SWC core versions with version identifiers in this plugin. What if you want to fix a bug or add a feature (for example, #9445)? If you do want to maintain support for different SWC core versions, I think it'd be better and just as much of a mess to build and include multiple versions of `swc_plugin.wasm` in each release of the plugin (so you'd have e.g. `swc_plugin-0.78.wasm`, `swc_plugin-0.76.wasm` for however many versions you'd decide to support).
- This would move version select to where you configure SWC plugins, instead of package.json. Next.js allows you to specify "an npm module package name or an absolute path to the `.wasm` binary itself" (source: https://nextjs.org/docs/architecture/nextjs-compiler#swc-plugins-experimental), so I'm admittedly not sure if this is ergonomic at all.
- My personal opinion is that you don't need to support anything but the latest version of SWC core. This _is_ an experimental feature, so I don't think anyone can expect backwards compatibility.

**Comment by victorandree** [comment]
@CaptainN To be fair, there should also be a (stable) Babel Plugin that achieves the same results, see <https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#reducing-bundle-size>. This is about the SWC plugin specifically.

**Comment by CaptainN** [comment]
@victorandree That's true, but it's pretty challenging to get that to work with nextjs, which uses SWC by default. (It has a way to change back to babel, but it breaks a bunch of stuff - not easy to pull off.)

**Comment by CaptainN** [comment]
How about if we just don't generate the giant file to start with - it's relatively simple to manually edit it so that you don't get that: https://github.com/dotansimha/graphql-code-generator/issues/9988

**Comment by victorandree** [comment]
For anyone still following this thread:

- I've released a new version of my forked plugin (https://www.npmjs.com/package/@victorandree/graphql-codegen-client-preset-swc-plugin/v/0.5.0) that uses `swc_core` v3.0.2 which makes it compatible with Next.js 15.
- I've tested that without the plugin, the `documents` map is present in `gql.ts`, and that activating it removes it. That is, the plugin is still relevant and still works.
- `swc_core` compatibility information is now available at: https://plugins.swc.rs.

Please note that I just make the bare minimum changes to make the plugin "work" for me. I don't know how newer versions work with the issues mentioned at the top of the issue, or with non-Next.js runtimes, or how to resolve the "version selection" problem.

I don't know if `swc_core` being on a non-0.x version implies anything about compatibility.

**Comment by karlhorky** [comment]
@YassinEldeeb if I understand correctly, there's no actionable way forward here?

Should [the `@graphql-codegen/client-preset-swc-plugin` package](https://www.npmjs.com/package/@graphql-codegen/client-preset-swc-plugin) be deprecated?

cc @dotansimha @enisdenjo

**Comment by eddeee888** [comment]
The plugin is moved to https://github.com/swc-project/plugins/tree/main/contrib/graphql-codegen-client-preset
So this PR is no longer required in this repo 🙂
